src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IsPatternOperator.cs

```diff
@@ -15,13 +15,126 @@ internal sealed partial class LocalRewriter
         {
-            var isPatternRewriter = new IsPatternExpressionLocalRewriter(node.Syntax, this);
-            BoundExpression result = isPatternRewriter.LowerIsPattern(node, node.Pattern, this._compilation, this._diagnostics);
-            isPatternRewriter.Free();
-            return result;
+            if (canProduceLinearSequence(node.DecisionDag.RootNode, whenTrueLabel: node.WhenTrueLabel, whenFalseLabel: node.WhenFalseLabel))
+            {
+                // If we can build a linear test sequence `(e1 && e2 && e3)` for the dag, do so.
+                var isPatternRewriter = new IsPatternExpressionLinearLocalRewriter(node, this);
+                BoundExpression result = isPatternRewriter.LowerIsPatternAsLinearTestSequence(node, whenTrueLabel: node.WhenTrueLabel, whenFalseLabel: node.WhenFalseLabel);
+                isPatternRewriter.Free();
+                return result;
+            }
+            else if (canProduceLinearSequence(node.DecisionDag.RootNode, whenTrueLabel: node.WhenFalseLabel, whenFalseLabel: node.WhenTrueLabel))
+            {
+                // If we can build a linear test sequence with the whenTrue and whenFalse labels swapped, then negate the
+                // result.  This would typically arise when the source contains `e is not pattern`.
+                var isPatternRewriter = new IsPatternExpressionLinearLocalRewriter(node, this);
+                BoundExpression result = isPatternRewriter.LowerIsPatternAsLinearTestSequence(node, whenTrueLabel: node.WhenFalseLabel, whenFalseLabel: node.WhenTrueLabel);
+                result = this._factory.Not(result);
+                isPatternRewriter.Free();
+                return result;
+            }
+            else
+            {
+                // We need to lower a generalized dag, so we produce a label for the true and false branches and assign to a temporary containing the result.
+                var isPatternRewriter = new IsPatternExpressionGeneralLocalRewriter(node.Syntax, this);
+                BoundExpression result = isPatternRewriter.LowerGeneralIsPattern(node);
+                isPatternRewriter.Free();
+                return result;
+            }
+
+            // Can the given decision dag node, and its successors, be generated as a sequence of
+            // linear tests with a single "golden" path to the try label and all other paths leading
+            // to the false label?  This occurs with an is-pattern expression that uses no "or" or "not"
+            // pattern forms.
+            static bool canProduceLinearSequence(
+                BoundDecisionDagNode node,
+                LabelSymbol whenTrueLabel,
+                LabelSymbol whenFalseLabel)
+            {
+                while (true)
+                {
+                    switch (node)
+                    {
+                        case BoundWhenDecisionDagNode w:
+                            Debug.Assert(w.WhenFalse is null);
```

Comment:
Why do we know this is always the case?

Reply:
Because is-pattern expressions don't have a when clause.

---
In reply to: [389167183](https://github.com/dotnet/roslyn/pull/42109#discussion_r389167183) [](ancestors = 389167183)

Reply:
Ah, right, I forgot this code would only be used for `is`